### PR TITLE
CEDS-1840 Add departure time question

### PR DESCRIPTION
**Pull request by bambuchaAdm**

### CODE REVIEW

File: app/forms/common/Time.scala [hunk 1]
```diff
@@ -51,11 +40,23 @@ object Time {
...
-      hourKey -> optional(text().verifying("dateTime.time.hour.error", correctHour))
-        .verifying("dateTime.time.hour.empty", _.nonEmpty),
-      minuteKey -> optional(text().verifying("dateTime.time.minute.error", correctMinute))
-        .verifying("dateTime.time.minute.empty", _.nonEmpty)
-    )(Time.apply)(Time.unapply)
+  val mapping: Mapping[Time] = {
+    def bind(hour: Option[String], minutes: Option[String]): Time = {
+      (hour, minutes) match {
+        case (Some(h), Some(m)) => Time(LocalTime.of(h.toInt, m.toInt))
+        case _ => throw new IllegalArgumentException("Could not build time - missing one of parameters")
```
Comment (PatrykRudnicki) [inline app/forms/common/Time.scala]: Same question like above

File: app/forms/DepartureDetails.scala [hunk 2]
```diff
@@ -16,18 +16,27 @@
...
+
+import forms.common.{Date, Time}
+import play.api.data.Forms._
+import play.api.data._
 import play.api.libs.json.{Json, OFormat}
 
-case class DepartureDetails(dateOfDeparture: Date) {
+case class DepartureDetails(dateOfDeparture: Date, timeOfDeparture: Time) {
+
+  def moment: LocalDateTime = LocalDateTime.of(dateOfDeparture.asLocalDate, timeOfDeparture.time)
```
Comment (MaciejRewera) [inline app/forms/DepartureDetails.scala]: Could you provide some more explanatory name? Without a context it is hard to say what it actually does.

File: app/forms/common/Date.scala [hunk 3]
```diff
@@ -25,25 +25,23 @@ import play.api.libs.json.{Json, OFormat}
...
+    date.atTime(LocalTime.of(0, 0, 0)).format(DateTimeFormatter.ISO_DATE_TIME)
 
   }
 
-  override def toString: String = LocalDate.of(year.getOrElse(0), month.getOrElse(0), day.getOrElse(0)).toString
-  def toLocalDateTimeString: String =
-    LocalDate.of(year.getOrElse(0), month.getOrElse(0), day.getOrElse(0)).atStartOfDay().toString + ":00"
+  override def toString: String = date.toString
+
+  def asLocalDate: LocalDate = date
```
[the remark this answers is not in the document]
Comment (bambuchaAdm) [reply app/forms/common/Date.scala]: That should be inlined

File: app/forms/common/Date.scala [hunk 4]
```diff
@@ -57,16 +55,35 @@ object Date {
...
+        case (Some(d), Some(m), Some(y)) => Try(LocalDate.of(y, m, d)).isSuccess
+        case _ => false
+      }
+    }
+
+    def bind(day: Option[Int], month : Option[Int], year: Option[Int]): Date = {
+      (day, month, year) match {
+        case (Some(d), Some(m), Some(y)) => Date(LocalDate.of(y, m, d))
+        case _ => throw new IllegalArgumentException("Could not bind local date when any is empty")
+      }
```
[the remark this answers is not in the document]
Comment (PatrykRudnicki) [reply app/forms/common/Date.scala]: Discussed, it's okay

File: app/forms/common/Time.scala (hunk 1 above)
Comment (PatrykRudnicki) [reply app/forms/common/Time.scala]: Disscussed, it's okay

File: app/forms/DepartureDetails.scala (hunk 2 above)
Comment (bambuchaAdm) [reply app/forms/DepartureDetails.scala]: departure moment  - as DepartureDetails#moment

File: test/helpers/FormMatchers.scala [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+ * limitations under the License.
+ */
+
+package helpers
+
+import org.scalatest.matchers.{BeMatcher, MatchResult}
+import play.api.data.{Form, FormError}
+
+trait FormMatchers {
+  val errorless: BeMatcher[Form[_]] = new BeMatcher[Form[_]] {
```
Comment (MaciejRewera) [inline test/helpers/FormMatchers.scala]: Nit: It took me a while to realise what it does. I would name sth like `withoutErrors`. But might be just a matter of slow thinking.

File: test/helpers/FormMatchers.scala [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+ * limitations under the License.
+ */
+
+package helpers
+
+import org.scalatest.matchers.{BeMatcher, MatchResult}
+import play.api.data.{Form, FormError}
+
+trait FormMatchers {
+  val withoutErrors: BeMatcher[Form[_]] = new BeMatcher[Form[_]] {
```
Comment (MaciejRewera) [inline test/helpers/FormMatchers.scala]: 👍

**Pull request merged**